operator/standalone/delete.go

```
@@ -0,0 +1,77 @@
+	pipeline "github.com/ccremer/go-command-pipeline"
+)
+
+const crossplaneManagedFinalizer = "finalizer.managedresource.crossplane.io"
+
+// DeleteStandalonePipeline is a pipeline that deletes an instance from the target deployment namespace.
+type DeleteStandalonePipeline struct {
+	client   client.Client
+	instance *v1alpha1.PostgresqlStandalone
+}
+
+// NewDeleteStandalonePipeline creates a new delete pipeline with the required dependencies.
+func NewDeleteStandalonePipeline(client client.Client, instance *v1alpha1.PostgresqlStandalone) *DeleteStandalonePipeline {
+	return &DeleteStandalonePipeline{
+		instance: instance,
+		client:   client,
+	}
+}
+
+// RunPipeline executes the pipeline with configured business logic steps.
+func (d *DeleteStandalonePipeline) RunPipeline(ctx context.Context) error {
+	return pipeline.NewPipeline().
+		WithSteps(
+			pipeline.NewStepFromFunc("delete helm release", d.deleteHelmRelease),
+			pipeline.NewStepFromFunc("delete namespace", d.deleteNamespace),
+			pipeline.NewStepFromFunc("delete connection secret", d.deleteConnectionSecret),
+		).
+		RunWithContext(ctx).Err()
+}
+
+// deleteHelmRelease removes the Helm Release from the cluster
+// The crossplaneManagedFinalizer is removed before the Release is deleted
+func (d *DeleteStandalonePipeline) deleteHelmRelease(ctx context.Context) error {
+	helmRelease := &helmv1beta1.Release{}
+	err := d.client.Get(ctx, client.ObjectKey{Name: d.instance.Status.HelmChart.DeploymentNamespace}, helmRelease)
+	if err != nil {
+		return err
+	}
+	removed := controllerutil.RemoveFinalizer(helmRelease, crossplaneManagedFinalizer)
+	if !removed {
+		return fmt.Errorf("cannot remove finalizer %s from helm release %s", crossplaneManagedFinalizer, helmRelease.Name)
+	}
+	// TODO add DeletionPropagation ???
+	return d.client.Delete(ctx, helmRelease)
+}
+
+// deleteNamespace removes the namespace of the PostgreSQL instance
+func (d *DeleteStandalonePipeline) deleteNamespace(ctx context.Context) error {
```

Do we want to return an error in case the namespace is somehow already deleted?

---

No, that's what we ultimately want :)

---

yep make sense